Change Get-FileHash to close file handles before writing output

HumanEquivalentUnit opened the pull request:
<!-- Anything that looks like this is a comment and can't be seen after the Pull Request is created. -->

# PR Summary

Fix #12473 where Get-FileHash holds the hashed file handles open unnecessarily long. 
Change is to hash the file contents, then dispose the file stream, then write the results, in that order.

## PR Context

`Get-ChildItem | Get-FileHash | Rename-Item -NewName {$_.Hash}` throws because the file is still open when trying to be renamed, this PR tries to fix that by closing the file handle before Get-FileHash writes output.

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.
- **[Breaking changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)**
    - [x] None
    - **OR**
    - [ ] [Experimental feature(s) needed](https://github.com/MicrosoftDocs/PowerShell-Docs/blob/staging/reference/6/Microsoft.PowerShell.Core/About/about_Experimental_Features.md)
        - [ ] Experimental feature name(s): <!-- Experimental feature name(s) here -->
- **User-facing changes**
    - [x] Not Applicable
    - **OR**
    - [ ] [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed: <!-- Number/link of that issue here -->
- **Testing - New and feature**
    - [ ] N/A or can only be tested interactively
    - **OR**
    - [x] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
- **Tooling**
    - [x] I have considered the user experience from a tooling perspective and don't believe tooling will be impacted.
    - **OR**
    - [ ] I have considered the user experience from a tooling perspective and enumerated concerns in the summary. This may include:
        - Impact on [PowerShell Editor Services](https://github.com/PowerShell/PowerShellEditorServices) which is used in the [PowerShell extension](https://github.com/PowerShell/vscode-powershell) for VSCode (which runs in a different PS Host).
        - Impact on Completions (both in the console and in editors) - one of PowerShell's most powerful features.
        - Impact on [PSScriptAnalyzer](https://github.com/PowerShell/PSScriptAnalyzer) (which provides linting & formatting in the editor extensions).
        - Impact on [EditorSyntax](https://github.com/PowerShell/EditorSyntax) (which provides syntax highlighting with in VSCode, GitHub, and many other editors).

HumanEquivalentUnit committed:
Add test for not holding the file open.

HumanEquivalentUnit committed:
Close file before writing output to pipeline.

Fix the use case `Get-FileHash | Rename-Item` 
by changing to `using () {}` and closing the hashed file before writing the output result.

msftclas commented:
[![CLA assistant check](https://cla.opensource.microsoft.com/pull/badge/not_signed)](https://cla.opensource.microsoft.com/PowerShell/PowerShell?pullRequest=12474) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.opensource.microsoft.com/PowerShell/PowerShell?pullRequest=12474) before we can accept your contribution.<br/><br/>:x: HumanEquivalentUnit [sign now](https://cla.opensource.microsoft.com/PowerShell/PowerShell?pullRequest=12474)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.opensource.microsoft.com/check/PowerShell/PowerShell?pullRequest=12474) it.</sub>

vexx32 commented on test/powershell/Modules/Microsoft.PowerShell.Utility/Get-FileHash.Tests.ps1:
```diff
@@ -77,4 +77,10 @@ Describe "Get-FileHash" -Tags "CI" {
+    
+    Context "Close file before writing hash details to pipeline" {
+        It "Should not hold the file open unnecessarily" {
+            {Get-FileHash $testDocument | Rename-Item -NewName {$_.Hash}} | Should -Not -Throw
```
Does Rename-Item throw a terminating error if it can't get at the file? Wondering if we need `-ErrorAction Stop` on this one.

We should probably add a comment here indicating that this changes the filename of `$testDocument` in case folks try to add further tests after this point using the same file. Alternatively, you could store the new file name and rename it back in an `AfterAll` block?

HumanEquivalentUnit replied:
Good points; I've added `-ErrorAction Stop`, and updated the test to make a copy of the document first, and remove it afterwards.

vexx32 reviewed (commented):
Looks good to me! Just a couple questions on the test. 😊

HumanEquivalentUnit committed:
Update close filehandle test

Copy the test document first, then remove the copy after, so it won't affect other tests.
Assumes the test drive is writable.

HumanEquivalentUnit committed:
Tidy up comment

vexx32 reviewed (approved):
Looks good, nice work! :blush:

vexx32 commented:
@PoshChan  please retry macos

PoshChan commented:
@vexx32, successfully started retry of `PowerShell-CI-macOS`

HumanEquivalentUnit commented:
NB. The failing `PowerShell-CI-static-analysis` test is about URLs in Markdown in the PowerShell 6 CHANGELOG, and is unrelated to these changes.

iSazonov commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/GetHash.cs:
```diff
@@ -128,14 +128,15 @@ protected override void ProcessRecord()
-                    bytehash = hasher.ComputeHash(openfilestream);
-
-                    hash = BitConverter.ToString(bytehash).Replace("-", string.Empty);
+                    using (Stream openfilestream = File.OpenRead(path))
+                    {
+                        bytehash = hasher.ComputeHash(openfilestream);
+                        hash = BitConverter.ToString(bytehash).Replace("-", string.Empty);
+                    }
+                    
                     WriteHashResult(Algorithm, hash, path);
```
We can not use `using` and put the write in`finally` like 
```c#
if (hash != null)
{                
    WriteHashResult(Algorithm, hash, path);
}
```

vexx32 replied:
We can, sure. Is there much of a reason to use that pattern? It requires an additional check which we don't really have a need to do.

HumanEquivalentUnit replied:
Is `using` something to avoid? I like how `using` keeps the scope for the open file together in just a few lines, but your suggestion would be fine as well - I can change it?

Or add another `openfilestream?.Dispose()` just before WriteHashResult, and keep the one in `finally` as well in case of exceptions?

iSazonov replied:
`using` is internally try-finally. Really it masks logic error. If we want to add something in the future, the proposed separation of logic will save us from possible problems like this.

vexx32 replied:
What logic error are you seeing here? What'd I miss? 😄

iSazonov replied:
Ideally it could be:
```c#
if (ComputeFileHash(Algorithm, path, out hash)
{
    WriteHashResult(Algorithm, hash, path);
}

private static bool ComputeFileHash(...)
{
    try
    {
    }
    catch
    {
    }
    finally
    {
    }
}

iSazonov replied:
> What logic error

This is the bug we fix here - we should dispose before write.

vexx32 replied:
That'll have the same issue, though.

This problem arises because we're not disposing the file handle before we write the result. That's what we need to do here, and separating them out like that will bring the problem back. 

We need to:

1. Open file
2. Read file
3. Compute hash
4. Close file handle
5. Write hash result

In that order, or we come back to the same problem here.

iSazonov replied:
In general, nested try-try (that `using` is internally) should be avoided until it is inevitable. This complicates the reading and understanding of the code. My second example is perfectly simple from this point of view.

HumanEquivalentUnit replied:
OK, code rewritten with dedicated ComputeHash method.

HumanEquivalentUnit committed:
Move work into dedicated ComputeFileHash method

HumanEquivalentUnit committed:
Fix CodeFactor reports.

Add more documentation comments.
Remove Algorithm argument from ComputeFileHash method.

HumanEquivalentUnit committed:
Fix CodeFactor reports.

Add param descriptions.

iSazonov commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/GetHash.cs:
```diff
@@ -123,49 +123,16 @@ protected override void ProcessRecord()
                     break;
             }
-
+            
```
Please revert. Below too.
Please turn on removing tail spaces in your editor.

HumanEquivalentUnit replied:
Please revert which parts, and rever them to what state?
I will check for tail spaces.

iSazonov replied:
My request is only for tail spaces. Editor can automatically do this for you.

HumanEquivalentUnit replied:
OK.
I'm using the GitHub web editor, it can't automatically remove tail spaces. I was only making a quick edit to add `using` and was not setup for a larger refactor. I think I have removed the tail spaces.

iSazonov commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/GetHash.cs:
```diff
@@ -123,52 +123,16 @@ protected override void ProcessRecord()
                 string hash = null;
-                Stream openfilestream = null;
-
-                try
+                
+                if (ComputeFileHash(path, out hash))
```
Please use modern syntax:
```c#
                if (ComputeFileHash(path, out string hash))
```

HumanEquivalentUnit replied:
The term "modern syntax" does not mean anything to me - which part is not modern, and what is the modern version?

iSazonov replied:
C# 8.0 allow inline variable definition. So we can remove `string hash = null;` and use `if (ComputeFileHash(path, out string hash))`

HumanEquivalentUnit replied:
I understand, thank you; code is updated.

HumanEquivalentUnit committed:
Fix ComputeFileHash method bugs

Move bytehash var into it.
Update how it returns success/failure state.

iSazonov commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/GetHash.cs:
```diff
@@ -190,6 +153,64 @@ protected override void EndProcessing()
+                openfilestream?.Dispose();
+            }
+            
+            return hashComputedSuccessfully;
```
You could do:
```suggestion
            return hash != null;
```

HumanEquivalentUnit replied:
I like that; changed and updated.

iSazonov commented:
@HumanEquivalentUnit Please rebase to get a fix for CI-static.

HumanEquivalentUnit committed:
Simplify ComputeHash return value

Remove hashCompletedSuccessfully variable and replace with a return value calculated from hash variable.

HumanEquivalentUnit committed:
Fix CodeFactor report.

Remove blank line before closing brace.

iSazonov commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/GetHash.cs:
```diff
@@ -190,6 +153,61 @@ protected override void EndProcessing()
+            byte[] bytehash = null;
+            Stream openfilestream = null;
+
+            hash = null;
```
If compiler does not ask about this explicit initialization please remove.

HumanEquivalentUnit replied:
If I remove it in a test, when I get to `return hash != null` then it says:

CS0269 Use of unassigned out parameter 'hash'

CS0177 The out parameter 'hash' must be assigned to before control leaves the current method

so I think it is needed.

iSazonov commented on test/powershell/Modules/Microsoft.PowerShell.Utility/Get-FileHash.Tests.ps1:
```diff
@@ -77,4 +77,17 @@ Describe "Get-FileHash" -Tags "CI" {
+            $testDocumentCopy = "${testDocument}-copy"
+            Copy-Item -Path $testdocument -Destination $testDocumentCopy
+            
+            {Get-FileHash -Path $testDocumentCopy | Rename-Item -NewName {$_.Hash} -ErrorAction Stop} | Should -Not -Throw
```
Positive check is better because we don't expect throw but a result:
```suggestion
            { Get-FileHash -Path $testDocumentCopy | Rename-Item -NewName { $_.Hash } -PassThru } | Should -Exist
```

HumanEquivalentUnit replied:
Good idea! I have updated it with a positive test using `Test-Path` to look for the renamed file, and using `-ErrorAction SilentlyContinue` so if the rename fails, the tests will behave predictably.

HumanEquivalentUnit committed:
Inline the hash variable declaration.

HumanEquivalentUnit committed:
Improve the file handle being held open tests

Change the test logic to try the rename, and then Test-Path the expected new name.
Make the context and description clearer.

HumanEquivalentUnit committed:
Merge remote-tracking branch 'upstream/master'

HumanEquivalentUnit commented:
> @HumanEquivalentUnit Please rebase to get a fix for CI-static.

OK, I followed https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/syncing-a-fork and think it has merged the updated markdown files. Fingers crossed.

HumanEquivalentUnit committed:
Fix bug Test-Path on the wrong path

HumanEquivalentUnit commented:
`PowerShell-CI-macos` check failure is a traceroute one, and not related to this PR; could one of you re-run it? Not sure I can.

@PoshChan please retry macos

iSazonov commented on test/powershell/Modules/Microsoft.PowerShell.Utility/Get-FileHash.Tests.ps1:
```diff
@@ -77,4 +77,19 @@ Describe "Get-FileHash" -Tags "CI" {
+            $testDocumentCopy = "${testDocument}-copy"
+            Copy-Item -Path $testdocument -Destination $testDocumentCopy
+
+            $newPath = Get-FileHash -Path $testDocumentCopy | Rename-Item -NewName {$_.Hash} -PassThru -ErrorAction SilentlyContinue
+            Test-Path -Path $newPath.FullName -ErrorAction SilentlyContinue | Should -Be $true
+
+            Remove-Item -Path $testDocumentCopy -ErrorAction SilentlyContinue
```
We shouldn't use `SilentlyContinue` without explicit needs. This can hide an error useful for troublethooting. Also we can use `-BeTrue`.
```suggestion
            $newPath = Get-FileHash -Path $testDocumentCopy | Rename-Item -NewName {$_.Hash} -PassThru
            Test-Path -Path $newPath.FullName | Should -BeTrue

            Remove-Item -Path $testDocumentCopy -Force
```

vexx32 replied:
We can skip `Test-Path` and just use `$Path | Should -Exist` too.

HumanEquivalentUnit replied:
That makes sense about SilentlyContinue - if those throw exceptions, something is going wrong which needs addressing; both suggestions added, thanks :)

HumanEquivalentUnit committed:
Remove SilentlyContinue from file close test

Co-Authored-By: Ilya <darpa@yandex.ru>

HumanEquivalentUnit committed:
Update path test to use Should -Exist

vexx32 commented:
@PoshChan please retry macos

PoshChan commented:
@vexx32, successfully started retry of `PowerShell-CI-macOS`

HumanEquivalentUnit commented:
Thank you for reviewing and improving it, @iSazonov !
